src/rt/dwarfeh.d

```diff
@@ -780,3 +807,99 @@ void terminate(uint line) @nogc
     printf("dwarfeh(%u) fatal error\n", line);
     abort();     // unceremoniously exit
 }
+
+
+/****************************** C++ Support *****************************/
+
+enum _Unwind_Exception_Class cppExceptionClass =
+        (cast(_Unwind_Exception_Class)'G' << 56) |
+        (cast(_Unwind_Exception_Class)'N' << 48) |
+        (cast(_Unwind_Exception_Class)'U' << 40) |
+        (cast(_Unwind_Exception_Class)'C' << 32) |
+        (cast(_Unwind_Exception_Class)'C' << 24) |
+        (cast(_Unwind_Exception_Class)'+' << 16) |
+        (cast(_Unwind_Exception_Class)'+' <<  8) |
+        (cast(_Unwind_Exception_Class)0 <<  0);
+
+enum _Unwind_Exception_Class cppExceptionClass1 = cppExceptionClass + 1;
+
+
+/*****************************************
+ * Get Pointer to Thrown Object if type of thrown object is implicitly
+ * convertible to the catch type.
+ * Params:
+ *      exceptionObject = language specific exception information
+ *      sti = type of catch clause
+ * Returns:
+ *      null if not caught, pointer to thrown object if caught
+ */
+void* getCppPtrToThrownObject(_Unwind_Exception* exceptionObject, CppTypeInfo sti)
+{
+    void* p;    // pointer to thrown object
+    if (exceptionObject.exception_class & 1)
+        p = CppExceptionHeader.toExceptionHeader(exceptionObject).ptr;
+    else
+        p = cast(void*)(exceptionObject + 1);           // thrown object is immediately after it
+
+    const tt = (cast(CppExceptionHeader*)p - 1).typeinfo;
+
+    if (tt.__is_pointer_p())
+        p = *cast(void**)p;
```

Comment:
Why do you do it here?
If C++ code throws pointer to a class, you should do upcast before (via `__do_catch`).

```

try
{
    throwCppException(); //throw new std::bad_exception();
}
catch(std.exception e) // we need to cast std::bad_exception** to std::exception** before the exception pointer dereferencing.
{
}
```